language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -118,6 +119,19 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         existing_borrows.is_empty()
     }
 
+    fn is_readable_reference(
```

Comment:
Nice!

I wonder if there is some way to extend this to also capture the check going on in `ImmBorrowField` since they are *super* close to one another

Reply:
Hmm, is_readable_reference already has a reference in hand (corresponding to the nonce that is being passed in) and it answers the question whether a read through the reference is possible.  In your scenario, you don't have a reference in hand but you are asking read borrow is possible.  freeze_ok does most of the hard work already.